File: C++/Baekjoon/Problem1753.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

int inf = 987654321;

vector<pair<int, int>> a[20001];
int d[20001];

void dijkstra(int start) {
    d[start] = 0;
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
    pq.push({0, start});

    while (!pq.empty()) {
        int cur = pq.top().second;
        int dis = pq.top().first;
        pq.pop();

        if (d[cur] < dis) continue;
        for (int i = 0; i < a[cur].size(); i++) {
            int next = a[cur][i].first;
            int nextDis = dis + a[cur][i].second;

            if (nextDis < d[next]) {
                d[next] = nextDis;
                pq.push({nextDis, next});
            }
        }
    }
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);

    int V, E, K;
    cin >> V >> E >> K;

    fill_n(d, 20001, inf);

    for (int i = 0; i < E; i++) {
        int u, v, w;
        cin >> u >> v >> w;
        a[u].push_back({v, w});
    }
    dijkstra(K);

    for (int i = 1; i <= V; i++) {
        if (d[i] == inf) cout << "INF\n";
        else cout << d[i] << "\n";
    }
}

File: Python/Baekjoon/Problem1753.py
import heapq

def dijkstra(start):
    d[start] = 0
    h = []
    heapq.heappush(h, (0, start))

    while h:
        dis, cur = heapq.heappop(h)
        if d[cur] < dis: continue
        for i in graph[cur]:
            nextDis = dis + i[1]
            if nextDis < d[i[0]]:
                d[i[0]] = nextDis
                heapq.heappush(h, (nextDis, i[0]))


V, E = map(int, input().split())
K = int(input())
inf = 987654321

d = [inf for i in range(V + 1)]
graph = [[] for i in range(V + 1)]

for i in range(E):
    a, b, c = map(int, input().split())
    graph[a].append((b, c))

dijkstra(K)

for i in range(1, V + 1):
    if d[i] == inf:
        print("INF")
    else:
        print(d[i])
